### ISSUE TITLE
sum performance

### ISSUE THREAD

**Issue by salbert83** [denizyuret/Knet.jl#271]
I've experienced this on Windows and Linux.  Here is an example

using Knet

function sum_test(n)
    a = rand(n)
    t = 0.0
    @time for i = 1:100 t += sum(a) end
    a_ = KnetArray(a)
    t = 0.0
    @time for i = 1:100 t += sum(a_) end
end

sum_test(10)
sum_test(10000)

------------------------------

With output
 0.000001 seconds
  0.006031 seconds
  0.000286 seconds
  0.006576 seconds

The Knet sum seems to scale well with the vector size, doing the computation in the CPU is still much faster.  Is there a better way to do this?

Thanks!

**Comment by denizyuret** [comment]
Reduction functions on GPU are notoriously slow.  @ddikbayir: you were
working on better algorithms, can you confirm these timings and see if your
algorithms work faster?

**Comment by ddikbayir** [comment]
My CUDA kernels are not generalized yet, meaning they work with specific multidimensional arrays and input sizes for now. That have been said, like @salbert83 pointed out, GPU sum in Knet scales well with the vector size. I measured the performance with larger vector sizes and starting from ~100K elements, GPU sum performs better. I also measured performance with NVIDIA optimized vector reduction kernel available in CUDA Samples ([CUDA Parallel Reduction](http://docs.nvidia.com/cuda/cuda-samples/index.html#cuda-parallel-reduction)). Again, vectors of very small sizes are reduced faster on CPU. So, we can conclude that if the vector size is not large enough to leverage the GPU advantages (large amount of cores etc.), the vector should be reduced on the CPU.
